### ISSUE TITLE
Stop gradient calculation for numeric stability trick in logsumexp

### ISSUE THREAD

**Issue by matt-gardner** [allenai/allennlp#635]
Tensorflow has code to stop gradients from propagating when you take a max for numeric stability reasons when computing logsumexp.  It'd be nice to have a similar behavior in our code.  I'm not sure if or how pytorch allows this kind of thing, though.

Relevant line: https://github.com/allenai/allennlp/blob/d47ac92f1f7fcd905618f4c928f5b95556a2997c/allennlp/nn/util.py#L554

Relevant line in tensorflow: https://github.com/tensorflow/tensorflow/blob/438604fc885208ee05f9eef2d0f2c630e1360a83/tensorflow/python/ops/math_ops.py#L1670-L1674

**Comment by matt-gardner** [comment]
This would just mean calling `.detach()`, I think.  I don't have a good handle on whether this would really affect anything, though.
